error[E0700]: hidden type for `impl Trait` captures lifetime that does not appear in bounds

With this seemingly ok code:

```rust
#![feature(async_await)]

use async_trait::async_trait;
use slog::Logger;

#[async_trait]
trait MyTrait {
    async fn myfn(&self, logger: Logger);
}

struct MyImpl;

#[async_trait]
impl MyTrait for MyImpl {
    async fn myfn(&self, _logger: Logger) {}
}
```

I get this odd error:

```
   Compiling redacted v0.1.0 (/redacted)
error[E0700]: hidden type for `impl Trait` captures lifetime that does not appear in bounds
  --> tests/test.rs:15:43
   |
15 |     async fn myfn(&self, _logger: Logger) {}
   |                                           ^
   |
note: hidden type `impl std::future::Future` captures the scope of call-site for function at 15:43
  --> tests/test.rs:15:43
   |
15 |     async fn myfn(&self, _logger: Logger) {}
   |                                           ^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0700`.
error: Could not compile `redacted`.

To learn more, run the command again with --verbose.
```

The error is caused by `slog::Logger`.

This seems to be a bug. As a workaround, I'm currently passing the logger in a wrapper-struct, but I hope to get this solved.

PS: I have no idea what's going on here, and how to simplify the example further.

--- a/src/expand.rs
+++ b/src/expand.rs
@@ -1,4 +1,4 @@
-use crate::lifetime::CollectLifetimes;
+use crate::lifetime::{has_async_lifetime, CollectLifetimes};
 use crate::parse::Item;
 use crate::receiver::{has_self_in_block, has_self_in_sig, ReplaceReceiver};
 use proc_macro2::{Span, TokenStream};
@@ -249,11 +249,13 @@ fn transform_block(context: Context, sig: &mut MethodSig, block: &mut Block, has
             .extend(where_clause.predicates);
     }
 
-    standalone
-        .decl
-        .generics
-        .params
-        .push(parse_quote!('async_trait));
+    if has_async_lifetime(&mut standalone, block) {
+        standalone
+            .decl
+            .generics
+            .params
+            .push(parse_quote!('async_trait));
+    }
 
     let mut types = standalone
         .decl

--- a/src/lifetime.rs
+++ b/src/lifetime.rs
@@ -1,6 +1,25 @@
 use proc_macro2::Span;
 use syn::visit_mut::{self, VisitMut};
-use syn::{ArgSelfRef, GenericArgument, Lifetime, TypeReference};
+use syn::{ArgSelfRef, Block, GenericArgument, Item, Lifetime, MethodSig, TypeReference};
+
+pub fn has_async_lifetime(sig: &mut MethodSig, block: &mut Block) -> bool {
+    let mut visitor = HasAsyncLifetime(false);
+    visitor.visit_method_sig_mut(sig);
+    visitor.visit_block_mut(block);
+    visitor.0
+}
+
+struct HasAsyncLifetime(bool);
+
+impl VisitMut for HasAsyncLifetime {
+    fn visit_lifetime_mut(&mut self, life: &mut Lifetime) {
+        self.0 |= life.to_string() == "'async_trait";
+    }
+
+    fn visit_item_mut(&mut self, _: &mut Item) {
+        // Do not recurse into nested items.
+    }
+}
 
 pub struct CollectLifetimes {
     pub lifetimes: Vec<Lifetime>,

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -176,3 +176,16 @@ mod issue11 {
         async fn example(self: Arc<Self>) {}
     }
 }
+
+// https://github.com/dtolnay/async-trait/issues/15
+mod issue15 {
+    use async_trait::async_trait;
+    use std::marker::PhantomData;
+
+    trait Trait {}
+
+    #[async_trait]
+    trait Issue15 {
+        async fn myfn(&self, _: PhantomData<dyn Trait + Send>) {}
+    }
+}

